nomos-services/data-availability/network/src/lib.rs

```diff
@@ -53,59 +62,88 @@ impl<B: NetworkBackend<RuntimeServiceId>, RuntimeServiceId> Debug
+    Membership: MembershipAdapter + Send + 'static,
 {
-    type Settings = NetworkConfig<B, RuntimeServiceId>;
-    type State = NetworkState<B, RuntimeServiceId>;
+    type Settings = NetworkConfig<Backend, RuntimeServiceId>;
+    type State = NetworkState<Backend, RuntimeServiceId>;
     type StateOperator = NoOperator<Self::State>;
-    type Message = DaNetworkMsg<B, RuntimeServiceId>;
+    type Message = DaNetworkMsg<Backend, RuntimeServiceId>;
 }
 
 #[async_trait]
-impl<B, RuntimeServiceId> ServiceCore<RuntimeServiceId> for NetworkService<B, RuntimeServiceId>
+impl<Backend, Membership, RuntimeServiceId> ServiceCore<RuntimeServiceId>
+    for NetworkService<Backend, Membership, RuntimeServiceId>
 where
-    B: NetworkBackend<RuntimeServiceId> + Send + 'static,
-    B::State: Send + Sync,
-    RuntimeServiceId: AsServiceId<Self> + Clone + Display + Send,
+    Backend: NetworkBackend<RuntimeServiceId> + Send + Sync + 'static,
+    Membership: MembershipAdapter + Send + Sync + 'static,
+    <Membership::MembershipService as ServiceData>::Message: 'static,
+    RuntimeServiceId: AsServiceId<Self>
+        + AsServiceId<Membership::MembershipService>
+        + Clone
+        + Display
+        + Send
+        + Sync
+        + Debug
+        + 'static,
 {
     fn init(
         service_state: OpaqueServiceStateHandle<Self, RuntimeServiceId>,
         _init_state: Self::State,
     ) -> Result<Self, overwatch::DynError> {
         Ok(Self {
-            backend: <B as NetworkBackend<RuntimeServiceId>>::new(
-                service_state.settings_reader.get_updated_settings().backend,
-                service_state.overwatch_handle.clone(),
-            ),
             service_state,
+            _phantom: PhantomData,
         })
     }
 
     async fn run(mut self) -> Result<(), overwatch::DynError> {
         let Self {
-            service_state:
-                OpaqueServiceStateHandle::<Self, RuntimeServiceId> {
-                    mut inbound_relay,
-                    lifecycle_handle,
-                    ..
-                },
-            mut backend,
+            mut service_state, ..
         } = self;
 
-        let mut lifecycle_stream = lifecycle_handle.message_stream();
+        let network_settings = service_state.settings_reader.get_updated_settings();
+
+        let mut backend = NetworkBackend::new(
+            network_settings.backend,
+            service_state.overwatch_handle.clone(),
+        );
+
+        let membership_relay = service_state
+            .overwatch_handle
+            .relay::<Membership::MembershipService>()
+            .await?;
+
+        let membership_adapter = Membership::new(membership_relay);
+
+        let mut stream = membership_adapter
+            .subscribe(ServiceType::DataAvailability)
+            .await
+            .map_err(|e| {
+                tracing::error!("Failed to subscribe to membership service: {:?}", e);
+                e
+            })?;
```

Comment:
This adapter could implement `MembershipHandler` as well I think

Reply:
MembershipHandler describes API on the subnetworks assigned membership snapshot.  
Adapter is used to subscribe to membership changes or get snapshot at (later for historical). When the snapshot is received we need to create new subnetwork assigned membership (da membership) for it. That is why there is one trait for MembershipAdapter which is used to subscribe to membership service, and another trait MembershipHandler used by da network components (behaviors) that use membership handler. 
Does that makes sense?